Feature: THREE.ShadowMapViewer

arya-s opened the pull request:
Hey,

as discussed in #5896, I created a simple utility to visualise a light's shadow map.

arya-s committed:
Created a simple utility for visualising a light's shadow map.
Added webgl example.
See issue #5896.

WestLangley commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,171 @@
+		height: 256,
+	};
+
+	var camera = new THREE.OrthographicCamera( window.innerWidth / - 2, window.innerWidth / 2, window.innerHeight / 2, window.innerHeight / - 2, -1, 1 );
```
To prevent user confusion, make the `near` plane positive so the frustum is in front of the camera. You will have to move the camera back a bit.

arya-s replied:
Thank you, I updated the code with your suggestions.

WestLangley commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,171 @@
+			//See: https://github.com/mrdoob/three.js/issues/5932
+			uniforms.tDiffuse.value = light.shadowMap;
+
+			renderer.autoClear = false; // To allow render overlay
```
It is probably not a good idea to change `autoClear` behind-the-back of the user. Please be sure to reset it to its prior value.

arya-s committed:
# This is a combination of 2 commits.
# The first commit's message is:

Changed ShadowMapViewer's camera to have a positive near plane, as per @WestLangley's suggestion.
Storing renderer.autoClear before setting it to false when rendering the ShadowMapViewer and then restoring the user's autoClear setting, as per @WestLangley's suggestion.
Added ShadowMapViewer.update(), needed when manipulating the ShadowMapViewer's size or dimension coordinates directly without calling .set() on either.
Changed renndering of the text label to not happen if a light's name is empty.
Changed size members from x,y to width, height.

# This is the 2nd commit message:

Forgot to finish comment.

bchevalier commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,190 @@
+					fragmentShader: shader.fragmentShader
+	} );
+	var plane = new THREE.PlaneBufferGeometry( frame.width, frame.height );
+	var mesh = new THREE.Mesh( plane, material );
```
Hi,
I was wondering, as what you are trying to implement is to render 2D textures (with no perspective) on the screen, could it be simpler to use the ```THREE.Sprite``` component instead of a ```THREE.Mesh```?

There is an example of such an implementation here: [webgl sprites in Threejs](https://github.com/mrdoob/three.js/blob/master/examples/webgl_sprites.html)

arya-s replied:
Hey,

that was actually the first thing I looked into. However, a light's shadowMap is a WebGLRenderTarget and THREE.Texture doesn't accept those. And a THREE.SpriteMaterial doesn't accept that either.

Finally, I think you actually have to render out the texture before you can access it.

Correct me if I'm wrong and missed something.

bchevalier replied:
I see. So basically, if we wanted this to work using ```THREE.Sprite```, ```THREE.SpriteMaterial ``` would have to accept ```THREE.WebGLRenderTarget```. I am not sure if it is feasible, but if it was done it could simplify a tiny bit the work of anyone building such HUDs.

mrdoob replied:
I wonder that too. I haven't looked at that code for a bit. Would it be hard to make `SpriteMaterial` support `WebGLRenderTarget`?

arya-s replied:
@mrdoob I haven't looked into it. But in case of ```light.shadowMap```, you have to render it first. Sprites and Textures don't seem to require such a pass.

I'm not sure about other WebGLRenderTargets though.

mrdoob replied:
> Edit: Maybe we should open a different issue and discuss that there?

Agreed!

WestLangley commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,190 @@
+	};
+
+	var camera = new THREE.OrthographicCamera( window.innerWidth / - 2, window.innerWidth / 2, window.innerHeight / 2, window.innerHeight / - 2, 1, 10 );
+	camera.position.set(0, 0, 1);
```
This is dangerous. The near plane passes through the world origin. You are placing the mesh at the origin.

    camera.position.set( 0, 0, 2 );

Also, consider checking your formatting. http://zz85.github.io/mrdoobapproves/

arya-s replied:
@WestLangley thanks, I overlooked that.

arya-s committed:
Moved ShadowMapViewer's camera back so that its near plane is not passing through the world origin, as per @WestLangley's suggestion.
Fixed some code formatting.

mrdoob commented:
Thanks!

WestLangley commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,190 @@
+			//always end up with the scene's first added shadow casting light's shadowMap
+			//in the shader
+			//See: https://github.com/mrdoob/three.js/issues/5932
+			uniforms.tDiffuse.value = light.shadowMap;
```
Hmmm. `light` is defined???. This does appear to work, but is it guaranteed to?

`light` is passed into the constructor as an argument only.

arya-s replied:
What do you mean by guaranteed to work?

If ```light``` is undefined, the whole thing fails. Which is the same behaviour as for example ```THREE.DirectionalLightHelper```.

WestLangley replied:
1. Where is the variable `light` set?
2. Is is guaranteed to always be defined?

arya-s replied:
1. Wherever the user defines his light setup, then passed in as an argument to the constructor only.
2. No.

I'm not sure about your concern here. IF ```light``` becomes undefined, the entire thing breaks down, similar to other utilities in three.js. Should the user be just warned and the application not break?

What would you suggest here?

WestLangley replied:
Sorry, I must not be explaining myself well. I'll try again. : - )

Look at `SpotLightHelper`:

    THREE.SpotLightHelper = function ( light ) {

	    THREE.Object3D.call( this );

	    this.light = light;

	    ...

	}

Note that `light` is retained as the property `this.light`.

Then, in `THREE.SpotLightHelper.prototype.update()`, it references `this.light`.

You do not do that.

Instead, you reference `light` in your `render()` method. Nowhere in your `render()` method do you define what `light` is.

It is as if a closure has been created -- which I expect it has, because it is working.

/ping @mrdoob -- who might be able to shed some light on this.

arya-s replied:
Well the constructor function encloses everything, and that's where ```light``` is an argument, so ```render()``` has access to that same argument.

WestLangley replied:
Yes. And your `render()` method is not part of the prototype. That is the difference.

arya-s replied:
I didn't put the methods on the prototype because I didn't want to expose internals of ```THREE.ShadowMapViewer``` such as the ```camera```, ```scene```, ```light```, etc. .

But I'm still not getting your point. Are you wondering why ```render()``` has access to ```light```, or are you concerned about every instance of ```THREE.ShadowMapViewer``` having their own ```render()```, ```update()```, etc.?

WestLangley replied:
Sorry, I should have noticed the subtle differences in your approach, and avoided stirring the pot. It is clear to me now.

It would be nice if a user could create one Viewer and use it to render any light's shadow map, though. But, that is not that big of a deal.

Thank you for your contribution!

arya-s replied:
Yea, that's an interesting approach. It would also allow some easier placing of the shadow maps (for example in a grid). Less overhead too (just one scene, one camera, many meshes).

I just tried to go about this like a light helper, one per object, but I guess I mixed things up with not using prototypes and such.

WestLangley replied:
What I meant was, expose `this.light`. That way, a user can instantiate one Viewer, and re-assign a different light. It is up to you.

arya-s replied:
Ahhh, that's a good suggestion.

What's the procedure here? Do I update and do another pull request or should someone with rights just patch this in?

mrdoob replied:
I think I would keep the current approach.

@WestLangley as far as I remember, we used `this.light` because there was no other way. I prefer the approach here whenever possible. Even if limiting the usage to `new THREE.ShadowMapViewer( ... )` may be less memory efficient than other approaches it also reduces the code paths we need to be aware of.

WestLangley replied:
@mrdoob OK.

@arya-s Almost always, the contributor makes a new pull request. Even if it is simple.

WestLangley commented on examples/webgl_shadowmapviewer.html:
```diff
@@ -0,0 +1,241 @@
+        dirLight.shadowCameraLeft = -15;
+        dirLight.shadowCameraTop  = 15;
+        dirLight.shadowCameraBottom = -15;
+        spotLight.shadowDarkness = 0.5;
```
Should be `dirLight`.

arya-s replied:
Thanks.

mrdoob replied:
Already fixed :)

WestLangley commented on examples/js/utils/ShadowMapViewer.js:
```diff
@@ -0,0 +1,190 @@
+ *		light.name = 'Sun';
+ *
+ *	3) Create a shadow map viewer for that light and set its size and position optionally:
+ *		var shadowMapViewer = THREE.ShadowMapViewer( light );
```
new THREE.ShadowMapViewer( light );
